### ISSUE TITLE
GNS3 2.0.0b3 VMware passing traffic issue

### ISSUE THREAD

**Issue by jtorri** [GNS3/gns3-gui#1800]
Windows 10 Prof 64bit, VMWare Workstation 12.5.  IOU switches run in VMWare GNS3 VM 2.0dev7
GNS3 2.0.0b3
I believe this issue was covered in some detail in issues 1092, 1408, and 1521.  My best impression is it has understandably be put on the back burner.  The issue does not occur in GNS3 1.5.3 run with all executed locally on Ubuntu 16.01
The issue is an IOU switch will pass traffic if the source of the traffic is an IOU router but not if the source is a VMWare VM virtual machines, e.g., Windows 7 x64.  Is someone is able to have this work I would be most grateful if they would list the specifics of their implementation.  
I have the same results whether using Winpcap or Npcap.

Thanks again!

**Comment by noplay** [comment]
You have issue only with VMware?

**Comment by jtorri** [comment]
Windows 7 VM cannot communicate with RouterOutside.
![throughswitch](https://cloud.githubusercontent.com/assets/10285348/22179657/9063bb3c-e028-11e6-8c10-86350421df52.png)
Windows 7 VM can communicate with RouterOutside.
![straighttorouter](https://cloud.githubusercontent.com/assets/10285348/22179667/9f9e24b6-e028-11e6-90ad-e1a2c787d361.png)

**Comment by noplay** [comment]
And VMware to asav work?

**Comment by jtorri** [comment]
Do not understand the question?  Except for the Windows VM the entire topology is running inside a GNS3 VM build from GNS3.VM.VMware.Workstation.2.0.0b3.zip, the version of GNS3 VM supplied with GNS3 2.0.0beta3.  The ASAv object does work.  All traffic from objects within this topology to the "outside", Cloud1, (via my wired NIC on my PC running Windows) flows through the ASAv object.  
The issue I am describing has solely to do with the IOUL2 switch object, in this topology Switch1.  Traffic from a VMWare Virtual machine, this is case Windows7 running locally on my PC, will not pass through Switch1.  But, if I "connect" the Windows 7 virtual machine to an IOUL3 object, in this case RouterOutside, traffic is passed out through Cloud1.
Does this answer your question?

**Comment by noplay** [comment]
If you connect directly the ASAv to the Windows VM does it work?

**Comment by jtorri** [comment]
![win7toasav](https://cloud.githubusercontent.com/assets/10285348/22273151/c1459d7e-e26c-11e6-8206-c7dc55ea1a7f.png)
No traffic is being passed from the Windows7 VM to the ASAv951 object.  Does this answer your question?

Thanks.

**Comment by noplay** [comment]
Ok this mean the issue is VMware not IOU. Thanks

**Comment by jtorri** [comment]
Could this be related to ubridge functionality on Windows?  As I mentioned before and you likely know, running GNS3 1.5x on Ubuntu 16.01 64bit all objects work using the same topology.  Traffic passes from the Windows7 VM out through the cloud.  Perhaps it would be best for me to accept linux is the best platform for GNS3 and stop asking you folks to put limited resources on the limitations of Windows.  GNS3 is a great piece of software and all seems to work on linux.
Here's the topology on Ubuntu where all works.
![topologyinubuntu16-01](https://cloud.githubusercontent.com/assets/10285348/22290812/fa04a27c-e2d0-11e6-9417-2af41e2e87ce.png)

**Comment by noplay** [comment]
Linux will work better the system is open an nothing interact with the network without you.

The problem is VMware is a closed software, working differently on all OS. On Windows we get access to the network thanks a VMware engineer but the way it's work is obscure it's not impossible that a VMware upgrade break something.

The other possibility is a software (antivirus) on your computer prevent us to read packet from vmware. Often user don't know, we don't have errors, just nothing happen.

You try to turn off all security softwares?

**Comment by noplay** [comment]
Can you try to replace winpcap by npcap? https://github.com/nmap/npcap/releases/tag/v0.80

**Comment by jtorri** [comment]
Yes, I uninstalled Winpcap, rebooted the PC, installed Npcap from the GNS3 install application, rebooted the PC, ran the same GNS3 topology pictured above and the results were the same.  No traffic.
Interestingly the same topology works in GNS3 1.5.1.  The only related configuration option missing after 1.5.X is one where one has the option of specifying the use of ubridge on virtual machines run within a GNS3 topology.  This is option is no longer available in versions of GNS3 starting with 2.0.  I believe you mentioned starting with GNS3 2.0 all virtual VMs use only ubridge.
This is from GNS3 1.5.1
![ubridgeoption](https://cloud.githubusercontent.com/assets/10285348/22313307/45acd088-e32a-11e6-89fe-03949baf2bbc.png)
Thanks again. Likely I will just use linux rather than Windows to run GNS3.

**Comment by jtorri** [comment]
One more point that may be relevant.  All topologies, whether running GNS3.1.5.X or GNS3 2.0.X are using VMWare Workstation Pro 12 to run the GNS3 VM and, obviously, the Windows virtual machines.
Thanks!

**Comment by rdogaru** [comment]
I have the same problem, no traffic is passing between VMWare Workstation virtual machine and IOU switch. Its interesting that if you change the switch port in  to a routed port, traffic will pass.

**Comment by grossmj** [comment]
I suspect the problem is with winpcap/npcap. I thought I fixed that months
ago. I will investigate.

**Comment by noplay** [comment]
Perhaps we should apply the apple code to windows:
https://github.com/GNS3/ubridge/blob/master/nio_ethernet.c#L43

**Comment by noplay** [comment]
Hmm no it's not supported on Windows

**Comment by noplay** [comment]
I found a sample from the Npcap author:
https://github.com/hsluoyz/UserBridge/blob/master/UserBridge.c

We have some differences. 

ubridge:
```c
pcap_open(device, 65535,
       PCAP_OPENFLAG_PROMISCUOUS |
       PCAP_OPENFLAG_NOCAPTURE_LOCAL |
	   PCAP_OPENFLAG_MAX_RESPONSIVENESS |
	   PCAP_OPENFLAG_NOCAPTURE_RPCAP,
	   10, NULL, pcap_errbuf);
``` 

npcap sample:
```c
 pcap_open(d->name,						// name of the device
			65536,							// portion of the packet to capture. 
			// 65536 grants that the whole packet will be captured on every link layer.
			PCAP_OPENFLAG_PROMISCUOUS |	// flags. We specify that we don't want to capture loopback packets, and that the driver should deliver us the packets as fast as possible
			PCAP_OPENFLAG_NOCAPTURE_LOCAL |
			PCAP_OPENFLAG_MAX_RESPONSIVENESS,
			500,							// read timeout
			NULL,							// remote authentication
			errbuf							// error buffer
			)) == NULL)
```

**Comment by noplay** [comment]
I see multiple thing:
* We use 65535 as packet size, and they use 65536 (I found other sample on the net with the same value)
* We use PCAP_OPENFLAG_NOCAPTURE_RPCAP, I'm not sure it's usefull because we don't make remote capture
* We have a timeout of 10 when they use 500

**Comment by grossmj** [comment]
I have ran some tests with one IOS router connected to a VMware VM with a Ethernet switch between and I confirm the bug is present in versions 2.0.x and 1.5.3 but not in versions 1.5.2 and 1.5.1.

Somehow we broken something between 1.5.1 and 1.5.2. I suspect some back-ported code refactoring or something similar.

**Comment by grossmj** [comment]
@jtorri @rdogaru 

I think I found the issue. Could you test with the following uBridge executable?

https://drive.google.com/file/d/0B558b5_SCiAHVTZqUVB4WVpRR1k/view?usp=sharing

Please unzip this somewhere and change the path to ubridge.exe in the GNS3 preferences (Server section).

**Comment by jtorri** [comment]
Will do. Likely will be this evening. Thanks again for all your work!


Sent from my T-Mobile 4G LTE Device
-------- Original message --------From: Jeremy Grossmann <notifications@github.com> Date: 2/14/17  10:18 AM  (GMT-05:00) To: GNS3/gns3-gui <gns3-gui@noreply.github.com> Cc: jtorri <jtorri@windstream.net>, Mention <mention@noreply.github.com> Subject: Re: [GNS3/gns3-gui] GNS3 2.0.0b3 VMware passing traffic issue (#1800) 
@jtorri @rdogaru
I think I found the issue. Could you test with the following uBridge executable?
https://drive.google.com/file/d/0B558b5_SCiAHVTZqUVB4WVpRR1k/view?usp=sharing
Please unzip this somewhere and change the path to ubridge.exe in the GNS3 preferences (Server section).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/GNS3/gns3-gui","title":"GNS3/gns3-gui","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/GNS3/gns3-gui"}},"updates":{"snippets":[{"icon":"PERSON","message":"@grossmj in #1800: @jtorri @rdogaru \r\n\r\nI think I found the issue. Could you test with the following uBridge executable?\r\n\r\nhttps://drive.google.com/file/d/0B558b5_SCiAHVTZqUVB4WVpRR1k/view?usp=sharing\r\n\r\nPlease unzip this somewhere and change the path to ubridge.exe in the GNS3 preferences (Server section).\r\n"}],"action":{"name":"View Issue","url":"https://github.com/GNS3/gns3-gui/issues/1800#issuecomment-279736007"}}}

**Comment by rdogaru** [comment]
Great, it seems that is working. I will check more next days.


Thanks a lot for you work !

**Comment by jtorri** [comment]
GNS3 2.0.0b3.  The two new files do not seem to solve the switch not passing traffic issue.  But, as I have said before, versions of GNS3 application after 1.5.3 do not seem as stable as version 1.5.3, at least on Windows.  Thanks again!

**Comment by grossmj** [comment]
Thanks for testing. Do you have the issue in 1.5.3?

**Comment by jtorri** [comment]
The two new files solved the problem of IOU L2 not passing traffic in GNS3 1.5.3.  I mention that is a comment here earlier.  I included a picture of the topology for completeness.  Let me know if this does not answer your question.  Thanks!

**Comment by grossmj** [comment]
Ok, this was to confirm because I had the issue in 1.5.3 without the two
new files. I haven't tested 2.0b3 with these files yet.

Thanks for helping to solve this complex issue :)

**Comment by grossmj** [comment]
@jtorri

I've just tested with 2.0 beta 3 and the 2 files solved the L2 issue. Make
sure you have the correct path to ubridge.exe in the GNS3 preferences.

**Comment by jtorri** [comment]
GNS3 management console.
Running GNS3 version 2.0.0b3 on Windows (64-bit) with Python 3.5.2 Qt 5.7.1 and PyQt 5.7.1.

Adding the two files solved the L2 issue for me in GNS3 2.0.0b3 as well.  Thanks!
